### PR TITLE
refactor(common/core/web, web): target-agnostic key events

### PR DESCRIPTION
**Pull request by jahorton**
After some investigation and analysis, I noticed that in KMW 12 we started tracking the active output target as part of the key event object.  Storing it on the key event object would cause said objects to have a rather awkward construction process once Web OSK-Core stuff is well under way.

The issue:  the OSK itself won't track the active OutputTarget whatsoever, which makes it quite hard for the OSK to specify this property for the event.  The object would be near-fully constructed, with this one property `undefined` until much later with little in the object's design signaling the need to fill it in before passing it to the `KeyboardProcessor` part of Core.  (It's _just_ that single field - `Ltarg` - that's the issue.)

But... there's little need for it to _actually be_ on that object; keyboards receive the output target as a separate parameter already.  After adjustment of some method signatures, we can remove it from `KeyEvent` and still maintain access where needed, passing it as an extra parameter to relevant Core methods.  This avoids the delayed-construction problem and removes one of the "spaghetti strands" currently entangling our OSK code.

While it _is_ a bit awkward to change some of the method signatures of Core like this... fortunately, Core's API isn't exactly documented or in wide use yet.  Presenting something cleaner whenever we do properly document and promote its independent use would be wise.

### CODE REVIEW

File: common/core/web/keyboard-processor/src/text/defaultOutput.ts [hunk 1]
```diff
@@ -24,17 +24,17 @@ namespace com.keyman.text {
...
+      if((char = DefaultOutput.forSpecialEmulation(Lkc, ruleBehavior)) != null) {
         return char;
-      } else if(!isMnemonic && ((char = DefaultOutput.forNumpadKeys(Lkc)) != null)) {
+      } else if(!isMnemonic && ((char = DefaultOutput.forNumpadKeys(Lkc, ruleBehavior)) != null)) {
         return char;
-      } else if((char = DefaultOutput.forUnicodeKeynames(Lkc)) != null) {
+      } else if((char = DefaultOutput.forUnicodeKeynames(Lkc, ruleBehavior)) != null) {
         return char;
-      } else if((char = DefaultOutput.forBaseKeys(Lkc)) != null) {
+      } else if((char = DefaultOutput.forBaseKeys(Lkc, ruleBehavior)) != null) {
```
Comment (jahorton) [inline common/core/web/keyboard-processor/src/text/defaultOutput.ts:37]: The latter two already (optionally) took the second parameter, allowing the method to return warnings for select scenarios.  The problem was... we never actually provided that parameter in 14.0.  Oops!

File: common/core/web/input-processor/src/text/inputProcessor.ts [hunk 2]
```diff
@@ -58,12 +58,8 @@ namespace com.keyman.text {
      * @returns     {Object}             A RuleBehavior object describing the cumulative effects of
      *                                   all matched keyboard rules.
      */
-    processKeyEvent(keyEvent: KeyEvent): RuleBehavior {
+    processKeyEvent(keyEvent: KeyEvent, outputTarget: OutputTarget): RuleBehavior {
```
Comment (mcdurdin) [inline common/core/web/input-processor/src/text/inputProcessor.ts:61]: Please add the new parameter to the doc comment above (and fix the existing param name `e` in that comment while you are at it 😉)

File: common/core/web/keyboard-processor/src/text/keyboardProcessor.ts [hunk 3]
```diff
@@ -108,8 +108,7 @@ namespace com.keyman.text {
      * @param   {boolean} usingOSK
```
Comment (mcdurdin) [inline common/core/web/keyboard-processor/src/text/keyboardProcessor.ts:108]:
```suggestion
     * @param   {OutputTarget} outputTarget   ____________________
```

This suggestion needs some text filled in...

**Review by mcdurdin — APPROVED** [review]
I think this is okay. It'd be good to be iteratively improving the function header comments as we go; I noted a couple of places where parameters were old. But there are many functions with inadequate header comments or none at all. Particularly it is helpful to know if a given function is private to KMW or public (and therefore documented / contractual).

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.64-alpha](https://keyman.com/downloads/releases/alpha/15.0.64)

**Pull request merged**